File: leetcode/algorithms/66_plus_one/main.py
from typing import List


class PlusOne:
    """
    - Time Complexity: O(N)
    - Space Complexity: O(1)
    """
    def plusOne(self, digits: List[int]) -> List[int]:
        for i in range(len(digits) - 1, -1, -1):
            digits[i] += 1

            if digits[i] < 10:
                return digits

            digits[i] -= 10

        return [1] + ([0] * len(digits))


    # Solution
    """
    Solution 1
    - Time Complexity: O(N)
    - Space Complexity: O(1)
    """
    def solution1(self, digits: List[int]) -> List[int]:
        n = len(digits)

        for i in range(n - 1, -1, -1):
            if digits[i] < 9:
                digits[i] += 1
                return digits

            digits[i] = 0

        return [1] + digits

    """
    Solution 1
    - Time Complexity: O(N)
    - Space Complexity: O(N)
    """
    def solution2(self, digits: List[int]) -> List[int]:
        s = "".join(map(str, digits))

        num_plus_one_str = str(int(s) + 1)

        return [int(digit) for digit in num_plus_one_str]

File: leetcode/algorithms/66_plus_one/main.dart
class PlusOne {
  /**
   * Complexities:
   *   N - Length of `digits`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(N)
   */
  List<int> plusOne(List<int> digits) {
    for (int i = digits.length - 1; i >= 0; i--) {
      digits[i]++;

      if (digits[i] < 10) {
        return digits;
      }

      digits[i] -= 10;
    }

    return [1] + List.filled(digits.length, 0);
  }


  // Solution
  /**
   * Complexities:
   *   N - Length of `digits`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(1)
   */
  List<int> solution(List<int> digits) {
    int n = digits.length;

    for (int i = n - 1; i >= 0; i--) {
      if (digits[i] < 9) {
        digits[i]++;
        return digits;
      }

      digits[i] = 0;
    }

    digits.insert(0, 1);

    return digits;
  }
}
